iree/compiler/Dialect/Flow/Transforms/Passes.cpp

```diff
@@ -65,17 +65,18 @@ void buildFlowTransformPassPipeline(OpPassManager &passManager) {
   //----------------------------------------------------------------------------
   passManager.addPass(createCanonicalizerPass());
 
+  // Flatten structured control flow to our CFG.
+  passManager.addNestedPass<FuncOp>(mhlo::createLegalizeControlFlowPass());
+  passManager.addNestedPass<FuncOp>(createHLOPreprocessingPass());
+
   // Frontload linalg-on-tensors transformations and dispatch region creation.
   if (clEnableLinalgOnTensorsDispatch) {
+    passManager.addNestedPass<FuncOp>(createCanonicalizerPass());
```

Comment:
I would rather move this if branch down to line 175 and do 

```
if (!clEnableLinalgOnTensorsDispatch) {
  ... // whatever exists from line 175 - 240 (or 230?) 
} else {
  passManager.addNestedPass<FuncOp>(createCanonicalizerPass());
  addHLOToLinalgOnTensorsPasses(passManager);
  passManager.addNestedPass<FuncOp>(createDispatchLinalgOnTensorsPass(
      clLinalgOnTensorsTileSizes, clLinalgOnTensorsEnableFusion));
}

Reply:
Did that things (based on my understanding lines 175-240 for hlo/flow dispatch region creating) but not conversion to HAL is failing 
```
failed to legalize operation 'flow.dispatch.workgroups' that was explicitly marked illegal
```

Reply:
The HAL translate pass isn't invoked this way. something is missing is baked in one of these passes prepares the executable for translation .